### PR TITLE
Review PR #435: Worker join/leave notifies foreman

### PR DESCRIPTION
**Pull request by jmelloy**
## Summary of PR #435

PR #435 (now merged) adds worker lifecycle notifications so the foreman AI learns when workers come online or drop off. This review branch fixes two gaps found during the review.

## What PR #435 does

- **`handle_worker_register`** in `ws_handlers.py` now calls `_trigger_foreman` after updating repos, emitting `[worker-online] worker_id=<id> repos=<...> agent_count=<n>` so the foreman can assign queued tasks to the newly available worker.
- **`handle_worker_disconnect`** emits `[worker-offline] worker_id=<id> reason=shutdown` for clean shutdowns.
- The `finally` cleanup in `routes/websocket.py` batches all stale workers into a single `[worker-offline] reason=disconnect` trigger to avoid stampeding the foreman on mass-disconnect.
- Foreman system prompt updated with a "Reacting to worker lifecycle events" section.
- 3 new tests covering online, graceful offline, and abrupt-disconnect paths.

### 1. `_trigger_foreman` docstring missing new event types (`ws_handlers.py`)
The docstring enumerated the supported `event` values but `worker-online` and `worker-offline` were absent, making the vocabulary list incorrect for anyone reading it.

**Fix:** Added both to the docstring list.

### 2. Foreman prompt silent on batch-disconnect format (`foreman/prompt.py`)
The mass-disconnect path batches multiple workers into one trigger (one `[worker-offline]` line per worker, newline-separated). The prompt only described the single-worker case, so the foreman had no guidance on how to handle a message containing several lines.

**Fix:** Added a sentence to the `reason=disconnect` bullet explaining that a single message can contain multiple `[worker-offline]` lines and that each should be handled independently.

## Non-blocking observations (not fixed)

- **Re-registration fires `worker-online` again**: Every `worker-register` message triggers a foreman notification, including repo-update re-registrations. This is harmless (the foreman re-checks pending tasks) but slightly noisy. Fixing it would require tracking first-registration state per agent, which isn't worth the complexity.
- **`worker-online`/`worker-offline` not in CLAUDE.md protocol table**: The WS message-type table in CLAUDE.md lists `agent-state`, `task-assigned`, etc., but the new foreman trigger events aren't listed. These are internal trigger names, not wire messages, so this is a documentation judgment call.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/foreman/prompt.py [hunk 1]
```diff
@@ -100,6 +100,8 @@
   unexpectedly (crash or network failure). Escalate to the human if no other workers
   are available or if critical tasks were actively running on this worker. Use
   get_task_status to inspect affected tasks before deciding to reassign.
+  A single message may contain multiple `[worker-offline]` lines when several
```
Comment (jmelloy) [inline backend/foreman/prompt.py:103]: The guidance to 'handle each line independently' is correct, but it may be worth also reminding the foreman to deduplicate or batch its reassignment actions when many workers drop at once, since naive independent handling of N lines could still result in N separate (potentially conflicting) reassignment attempts.

File: backend/ws_handlers.py [hunk 2]
```diff
@@ -138,7 +138,8 @@ async def _trigger_foreman(
 
     ``event`` mirrors the plan's trigger-type vocabulary:
     ``chat``, ``task-complete``, ``followup-done``, ``needs-input``,
-    ``claude-auth``, ``periodic-check``.
+    ``claude-auth``, ``periodic-check``, ``worker-online``,
```
Comment (jmelloy) [inline backend/ws_handlers.py:141]: The updated docstring now lists `worker-online` and `worker-offline` but the broader docstring block (not shown) likely still does not document the expected payload format or what `event` is used for beyond this list. A brief note that these two values indicate worker lifecycle events (as opposed to task events) would make the distinction clearer for future contributors.

**Review by jmelloy — COMMENTED** [review]
- Both changes are purely documentation/comment fixes with no logic alterations, so there is no risk of behavioral regression.
- The docstring update in `ws_handlers.py` correctly adds the two new event types (`worker-online`, `worker-offline`) that PR #435 introduced, keeping the docstring in sync with actual usage.
- The prompt addition in `prompt.py` usefully clarifies that batched disconnect messages can carry multiple `[worker-offline]` lines, which is important for correct foreman behavior given the batching logic added in `routes/websocket.py`.
- No bugs, security issues, or design problems are introduced by these changes; both are low-risk, high-value documentation improvements.
- Consider whether the docstring should also note the new `worker-online` / `worker-offline` values in the `event` parameter description elsewhere if `_trigger_foreman` has a call-site or type annotation that lists allowed values.

**Review by jmelloy — COMMENTED** [review]
## Code Review: PR #438 — Fix docstring gap + clarify batch-disconnect prompt

**Verdict: Approve with minor notes.** Two targeted fixes, both correct. No functional change — purely documentation/prompt improvements.

---

### Overview

PR #435 (merged) added worker lifecycle notifications (`[worker-online]` / `[worker-offline]`) so the foreman AI learns when workers join or leave. PR #438 fixes two documentation gaps found during that review:

1. **`ws_handlers.py` docstring** — `_trigger_foreman`'s event vocabulary list was missing the two new event types.
2. **`foreman/prompt.py`** — The `reason=disconnect` bullet gave no guidance on the mass-disconnect case, where a single trigger can carry multiple `[worker-offline]` lines joined by `\n`.

---

### Analysis

#### `backend/ws_handlers.py` — docstring fix (lines 138–140)

```diff
-    ``claude-auth``, ``periodic-check``.
+    ``claude-auth``, ``periodic-check``, ``worker-online``,
+    ``worker-offline``.
```

Correct and complete. The list now matches all actual callers of `_trigger_foreman`. No functional impact.

#### `backend/foreman/prompt.py` — batch-disconnect clarification (lines 100–101)

```diff
+  A single message may contain multiple `[worker-offline]` lines when several
+  workers disconnect simultaneously — handle each line independently.
```

Accurate. The batch code in `routes/websocket.py` joins lines with `\n` and sends them as one trigger:

```python
offline_lines = "\n".join(
    f"[worker-offline] worker_id={wid} reason=disconnect"
    for wid in sorted(stale_worker_ids)
)
```

The prompt now correctly prepares the foreman for this format.

---

### Minor observations (non-blocking)

**Prompt guidance could be slightly more specific.**
"Handle each line independently" is correct but terse. The foreman already knows what to do for a single disconnect (check tasks, escalate if critical work was running); stating this explicitly — e.g., "apply the same assessment you would for a single-worker disconnect to each line" — would be more direct. Not a blocker; the existing bullet text above provides enough context for the model to infer the right behaviour.

**`worker-online`/`worker-offline` absent from CLAUDE.md protocol table.**
The WS message-type table documents wire messages, not internal trigger names, so omitting them is defensible. Still, a note in the table (even as `[foreman-trigger]` sub-types) would help future contributors. Low priority.

**Re-registration noise (carried forward from #435).**
Every `worker-register` fires `worker-online`, including routine re-registrations. This is called out in the PR description as a known, acceptable trade-off. Agreed — tracking first-registration state per agent is complexity without clear payoff.

---

### Test coverage

PR #435 added three tests covering online, graceful offline, and abrupt-disconnect paths. PR #438's changes are pure documentation/prompt text; no new tests needed. Existing tests continue to validate the functional behaviour.

---

### Summary

Both fixes are correct, minimal, and targeted. The prompt addition properly documents a real edge case (batch disconnect) that the foreman would otherwise have no guidance for. The docstring fix keeps the vocabulary list accurate for future maintainers. Ready to merge.
